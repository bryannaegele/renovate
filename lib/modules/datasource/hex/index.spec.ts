import { mockDeep } from 'jest-mock-extended';
import { GetPkgReleasesConfig, getPkgReleases } from '..';
import { Fixtures } from '../../../../test/fixtures';
import * as httpMock from '../../../../test/http-mock';
import { hostRules } from '../../../../test/util';
import { HexDatasource } from '.';

const tlsCertificateCheckAPIResponse = Fixtures.get(
  'tls_certificate_check.json',
);

const tlsCertificateCheckRegistryResponse = Fixtures.getBinary(
  'tls_certificate_check.bin.gz',
);

jest.mock('../../../util/host-rules', () => mockDeep());

const baseHexpmUrl = 'https://hex.pm/api';
const baseRegistryUrl = 'https://repo.hex.pm';
const privateBaseRegistryUrl = 'https://getoban.pro/repo';
const datasource = HexDatasource.id;
let config: GetPkgReleasesConfig;

describe('modules/datasource/hex/index', () => {
  beforeEach(() => {
    hostRules.hosts.mockReturnValue([]);
    hostRules.find.mockReturnValue({});
    config = {
      datasource,
      packageName: 'replace',
    };
  });

  describe('getReleases', () => {
    it('returns null for 404', async () => {
      httpMock.scope(baseRegistryUrl).get('/packages/some_package').reply(404);
      expect(
        await getPkgReleases({ ...config, packageName: 'some_package' }),
      ).toBeNull();
    });

    it('returns null for 401', async () => {
      httpMock.scope(baseRegistryUrl).get('/packages/some_package').reply(401);
      expect(
        await getPkgReleases({ ...config, packageName: 'some_package' }),
      ).toBeNull();
    });

    // do we have to throw for for >401 responses? Don't see this elsewhere
    // it('throws for 429', async () => {
    //   httpMock.scope(baseRegistryUrl).get('/packages/some_crate').reply(429);
    //   await expect(
    //     getPkgReleases({ datasource, packageName: 'some_crate' }),
    //   ).rejects.toThrow(EXTERNAL_HOST_ERROR);
    // });
    // it('throws for 429', async () => {
    //   httpMock.scope(baseUrl).get('/packages/some_crate').reply(429);
    //   await expect(
    //     getPkgReleases({ datasource, packageName: 'some_crate' }),
    //   ).rejects.toThrow(EXTERNAL_HOST_ERROR);
    // });
    // it('throws for 5xx', async () => {
    //   httpMock.scope(baseUrl).get('/packages/some_crate').reply(502);
    //   await expect(
    //     getPkgReleases({ datasource, packageName: 'some_crate' }),
    //   ).rejects.toThrow(EXTERNAL_HOST_ERROR);
    // });

    it('returns null for unknown error', async () => {
      httpMock
        .scope(baseRegistryUrl)
        .get('/packages/some_package')
        .replyWithError('');
      expect(
        await getPkgReleases({ ...config, packageName: 'some_package' }),
      ).toBeNull();
    });

    it('returns null with wrong auth token', async () => {
      httpMock
        .scope(baseRegistryUrl, {
          reqheaders: {
            authorization: 'abc',
          },
        })
        .get('/packages/tls_certificate_check')
        .reply(401);

      hostRules.find.mockReturnValueOnce({
        authType: 'Token-Only',
        token: 'abc',
      });

      const res = await getPkgReleases({
        ...config,
        packageName: 'tls_certificate_check',
      });

      expect(res).toBeNull();
    });

    it('processes real data', async () => {
      httpMock
        .scope(baseRegistryUrl)
        .get('/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckRegistryResponse);

      httpMock
        .scope(baseHexpmUrl)
        .get('/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckAPIResponse);

      const res = await getPkgReleases({
        ...config,
        packageName: 'tls_certificate_check',
      });
      expect(res).toMatchSnapshot();
      expect(res).not.toBeNull();
      expect(res).toBeDefined();
    });

    it('process public repo without auth', async () => {
      httpMock
        .scope(baseRegistryUrl)
        .get('/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckRegistryResponse);

      httpMock
        .scope(baseHexpmUrl)
        .get('/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckAPIResponse);

      hostRules.find.mockReturnValueOnce({});
      const res = await getPkgReleases({
        ...config,
        packageName: 'tls_certificate_check',
      });
      expect(res).toMatchSnapshot();
      expect(res).not.toBeNull();
      expect(res).toBeDefined();
    });

    it('extracts deprecated info', async () => {
      httpMock
        .scope(baseRegistryUrl)
        .get('/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckRegistryResponse);

      httpMock
        .scope(baseHexpmUrl)
        .get('/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckAPIResponse);

      hostRules.find.mockReturnValueOnce({});
      const res = await getPkgReleases({
        ...config,
        packageName: 'tls_certificate_check',
      });
      expect(res?.releases.some((rel) => rel.isDeprecated)).toBeTrue();
    });

    it('processes a private organization repo with auth', async () => {
      httpMock
        .scope(baseRegistryUrl, {
          reqheaders: {
            authorization: 'abc',
          },
        })
        .get('/repos/private_org/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckRegistryResponse);

      httpMock
        .scope(baseHexpmUrl, {
          reqheaders: {
            authorization: 'abc',
          },
        })
        .get('/repos/private_org/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckAPIResponse);

      hostRules.find.mockReturnValue({
        authType: 'Token-Only',
        token: 'abc',
      });
      const result = await getPkgReleases({
        ...config,
        packageName: 'org:private_org:tls_certificate_check',
      });

      expect(result).not.toBeNull();
    });

    // repo tests
    it('processes a private repo with auth', async () => {
      httpMock
        .scope(privateBaseRegistryUrl, {
          reqheaders: {
            authorization: 'abc',
          },
        })
        .get('/packages/tls_certificate_check')
        .reply(200, tlsCertificateCheckRegistryResponse);
      hostRules.find.mockReturnValueOnce({
        authType: 'Token-Only',
        token: 'abc',
      });
      const result = await getPkgReleases({
        ...config,
        packageName: 'repo:private_repo:tls_certificate_check',
        registryUrls: ['https://getoban.pro/repo'],
      });
      expect(result).not.toBeNull();
    });
  });
});
