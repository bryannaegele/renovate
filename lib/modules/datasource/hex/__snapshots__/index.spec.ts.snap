// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/datasource/hex/index getReleases process public repo without auth 1`] = `
{
  "changelogUrl": "https://github.com/g-andrade/tls_certificate_check/blob/master/CHANGELOG.md",
  "homepage": "https://hex.pm/packages/tls_certificate_check",
  "registryUrl": "https://repo.hex.pm",
  "releases": [
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-05-21T00:57:53.301Z",
      "version": "1.0.0-beta",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-05-21T01:02:49.271Z",
      "version": "1.0.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-05-21T09:56:01.525Z",
      "version": "1.0.1",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-10-16T18:05:41.581Z",
      "version": "1.0.2",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-12-05T19:33:37.712Z",
      "version": "1.1.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-12-08T20:54:17.217Z",
      "version": "1.1.1",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-03-12T22:07:36.263Z",
      "version": "1.2.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-04-02T15:42:35.731Z",
      "version": "1.3.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-04-16T23:00:02.486Z",
      "version": "1.4.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-05-13T18:25:16.258Z",
      "version": "1.5.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-05-30T19:57:29.487Z",
      "version": "1.6.0",
    },
    {
      "releaseTimestamp": "2021-07-08T22:09:21.845Z",
      "version": "1.7.0",
    },
    {
      "releaseTimestamp": "2021-08-31T20:17:24.102Z",
      "version": "1.8.0",
    },
    {
      "releaseTimestamp": "2021-09-03T13:57:22.713Z",
      "version": "1.9.0",
    },
    {
      "releaseTimestamp": "2021-10-01T18:01:02.256Z",
      "version": "1.10.0",
    },
    {
      "releaseTimestamp": "2021-10-28T17:38:29.649Z",
      "version": "1.11.0",
    },
    {
      "releaseTimestamp": "2022-02-02T16:40:07.824Z",
      "version": "1.12.0",
    },
    {
      "releaseTimestamp": "2022-03-21T16:46:32.005Z",
      "version": "1.13.0",
    },
    {
      "releaseTimestamp": "2022-04-27T14:12:18.990Z",
      "version": "1.14.0",
    },
    {
      "releaseTimestamp": "2022-07-20T16:08:34.371Z",
      "version": "1.15.0",
    },
    {
      "releaseTimestamp": "2022-10-11T10:52:36.095Z",
      "version": "1.16.0",
    },
    {
      "releaseTimestamp": "2023-01-11T18:42:45.919Z",
      "version": "1.17.0",
    },
    {
      "releaseTimestamp": "2023-01-12T01:15:19.695Z",
      "version": "1.17.1",
    },
    {
      "releaseTimestamp": "2023-01-12T16:36:16.103Z",
      "version": "1.17.2",
    },
    {
      "releaseTimestamp": "2023-01-17T15:12:42.271Z",
      "version": "1.17.3",
    },
    {
      "releaseTimestamp": "2023-02-19T19:23:26.806Z",
      "version": "1.17.4",
    },
    {
      "releaseTimestamp": "2023-03-20T18:07:34.489Z",
      "version": "1.18.0",
    },
    {
      "releaseTimestamp": "2023-05-01T22:56:38.169Z",
      "version": "1.18.1",
    },
    {
      "releaseTimestamp": "2023-05-30T10:51:23.377Z",
      "version": "1.19.0",
    },
    {
      "releaseTimestamp": "2023-08-22T10:56:44.511Z",
      "version": "1.20.0",
    },
    {
      "releaseTimestamp": "2023-12-12T16:32:01.017Z",
      "version": "1.21.0",
    },
    {
      "releaseTimestamp": "2024-03-16T15:52:29.677Z",
      "version": "1.22.0",
    },
    {
      "releaseTimestamp": "2024-03-16T16:21:24.461Z",
      "version": "1.22.1",
    },
  ],
  "sourceUrl": "https://github.com/g-andrade/tls_certificate_check",
}
`;

exports[`modules/datasource/hex/index getReleases processes real data 1`] = `
{
  "changelogUrl": "https://github.com/g-andrade/tls_certificate_check/blob/master/CHANGELOG.md",
  "homepage": "https://hex.pm/packages/tls_certificate_check",
  "registryUrl": "https://repo.hex.pm",
  "releases": [
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-05-21T00:57:53.301Z",
      "version": "1.0.0-beta",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-05-21T01:02:49.271Z",
      "version": "1.0.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-05-21T09:56:01.525Z",
      "version": "1.0.1",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-10-16T18:05:41.581Z",
      "version": "1.0.2",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-12-05T19:33:37.712Z",
      "version": "1.1.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2020-12-08T20:54:17.217Z",
      "version": "1.1.1",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-03-12T22:07:36.263Z",
      "version": "1.2.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-04-02T15:42:35.731Z",
      "version": "1.3.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-04-16T23:00:02.486Z",
      "version": "1.4.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-05-13T18:25:16.258Z",
      "version": "1.5.0",
    },
    {
      "isDeprecated": true,
      "releaseTimestamp": "2021-05-30T19:57:29.487Z",
      "version": "1.6.0",
    },
    {
      "releaseTimestamp": "2021-07-08T22:09:21.845Z",
      "version": "1.7.0",
    },
    {
      "releaseTimestamp": "2021-08-31T20:17:24.102Z",
      "version": "1.8.0",
    },
    {
      "releaseTimestamp": "2021-09-03T13:57:22.713Z",
      "version": "1.9.0",
    },
    {
      "releaseTimestamp": "2021-10-01T18:01:02.256Z",
      "version": "1.10.0",
    },
    {
      "releaseTimestamp": "2021-10-28T17:38:29.649Z",
      "version": "1.11.0",
    },
    {
      "releaseTimestamp": "2022-02-02T16:40:07.824Z",
      "version": "1.12.0",
    },
    {
      "releaseTimestamp": "2022-03-21T16:46:32.005Z",
      "version": "1.13.0",
    },
    {
      "releaseTimestamp": "2022-04-27T14:12:18.990Z",
      "version": "1.14.0",
    },
    {
      "releaseTimestamp": "2022-07-20T16:08:34.371Z",
      "version": "1.15.0",
    },
    {
      "releaseTimestamp": "2022-10-11T10:52:36.095Z",
      "version": "1.16.0",
    },
    {
      "releaseTimestamp": "2023-01-11T18:42:45.919Z",
      "version": "1.17.0",
    },
    {
      "releaseTimestamp": "2023-01-12T01:15:19.695Z",
      "version": "1.17.1",
    },
    {
      "releaseTimestamp": "2023-01-12T16:36:16.103Z",
      "version": "1.17.2",
    },
    {
      "releaseTimestamp": "2023-01-17T15:12:42.271Z",
      "version": "1.17.3",
    },
    {
      "releaseTimestamp": "2023-02-19T19:23:26.806Z",
      "version": "1.17.4",
    },
    {
      "releaseTimestamp": "2023-03-20T18:07:34.489Z",
      "version": "1.18.0",
    },
    {
      "releaseTimestamp": "2023-05-01T22:56:38.169Z",
      "version": "1.18.1",
    },
    {
      "releaseTimestamp": "2023-05-30T10:51:23.377Z",
      "version": "1.19.0",
    },
    {
      "releaseTimestamp": "2023-08-22T10:56:44.511Z",
      "version": "1.20.0",
    },
    {
      "releaseTimestamp": "2023-12-12T16:32:01.017Z",
      "version": "1.21.0",
    },
    {
      "releaseTimestamp": "2024-03-16T15:52:29.677Z",
      "version": "1.22.0",
    },
    {
      "releaseTimestamp": "2024-03-16T16:21:24.461Z",
      "version": "1.22.1",
    },
  ],
  "sourceUrl": "https://github.com/g-andrade/tls_certificate_check",
}
`;
